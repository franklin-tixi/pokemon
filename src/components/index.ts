/**
 * Barrel for components
 */

/** COMMON */
export * from "./AppCrudBar";

/** BUTTON */
export * from "./button/AppButton";

/** DIALOG */
export * from "./dialog/AppFormDialog";

/** INPUTS */
export * from "./inputs/AppTextField";

/** SLIDER */
export * from "./slider/AppSliderBar";

/** TABLE */
export * from "./table/AppTable";
