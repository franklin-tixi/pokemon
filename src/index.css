/********************************************************
*                           ROOT                        *
********************************************************/
:root {
  --main-color: #262626;
  --main-border-color: #b3b3b3;
  --main-border-color-primary: #197ccc;
  --main-border-radius: 3px;
  --main-bg-color: #fff;
  --main-bg-color-primary: #248ce4;
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;
  color: var(--main-color);
  background-color: var(--main-bg-color);
  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
  margin: 20px 40px 40px 40px;
}

/********************************************************
*                         TABLE                         *
********************************************************/
table,
tr,
td {
  border-collapse: collapse;
}

td {
  border-collapse: collapse;
  padding: 5px 40px;
  text-align: center;
}

thead {
  font-weight: 600;
}

/********************************************************
*                    BEAUTY SCROLL BAR                  *
********************************************************/
@media (min-width: 100px) {
  ::-webkit-scrollbar {
    width: 11px;
    height: 11px;
  }

  ::-webkit-scrollbar-track {
    background: rgba(224, 224, 224, 1);
  }

  ::-webkit-scrollbar-thumb {
    border-radius: 10px;
    background: #f5f5f5;
    background-color: #d9d9d9;
    box-shadow: inset 0 0 8px rgba(12, 9, 9, 0.5) !important;
  }
}

/********************************************************
*                         ICON                          *
********************************************************/
i.edit {
  background-image: url(../../../src/assets/images/icons/edit.png);
  background-repeat: no-repeat;
  background-size: 18px;
  background-position-x: 15%;
  background-position-y: 50%;
  padding-right: 20px;
  padding-left: 20px;
}

i.delete {
  background-image: url(../../../src/assets/images/icons/delete.png);
  background-repeat: no-repeat;
  background-size: 18px;
  background-position-x: 15%;
  background-position-y: 50%;
  padding-right: 20px;
  padding-left: 20px;
}

i:hover {
  cursor: pointer;
}
