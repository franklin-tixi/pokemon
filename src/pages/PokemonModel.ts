/**
 * Class model pokemon
 */
export class PokemonModel {
  public id: number = 0;
  public name: string = "";
  public image: string = "";
  public attack: number = 50;
  public defense: number = 50;
  public idAuthor: number = 1;
  public hp: number = 1;
  public type: string = "water";

  constructor(data?: Partial<PokemonModel>) {
    //@ts-ignore
    if (data?.id_author) {
      //@ts-ignore
      data.idAuthor = data.id_author;
      //@ts-ignore
      delete data.id_author;
    }
    Object.assign(this, data);
  }
}
