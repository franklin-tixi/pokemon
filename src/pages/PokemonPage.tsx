import { PokemonModel, PokemonService } from "@pages";
import {
  AppCrudBar,
  AppFormDialog,
  AppTable,
  AppTextField,
  AppSliderBar,
} from "@component";
import { CrudHelper, CrudState } from "@helpers/crud";

interface State extends CrudState<PokemonModel> {
  search: string;
}

/**
 * Class Page para Crud
 */
export class PokemonPage extends CrudHelper<PokemonModel, State> {
  /**
   * State of Pokemon
   */
  public state: State = {
    search: "",
    openForm: false,
    elementIndex: -1,
    element: new PokemonModel({}),
    list: [],
  };

  /********************************************************
   *                         METHODS                      *
   ********************************************************/
  //#region METHODS
  /**
   * Load pokemon data
   */
  async loadPokemons() {
    this.load(PokemonService.getAll);
  }

  /**
   * Submit in dialog
   */
  async submit() {
    if (this.state.element.id > 0) {
      this.editPokemon();
    } else {
      this.createPokemon();
    }
  }
  /**
   * Create pokemon
   */
  async createPokemon() {
    this.create(PokemonService.create, PokemonService.getAll);
  }

  /**
   * Edit pokemon
   */
  async editPokemon() {
    this.edit(PokemonService.edit);
  }

  /**
   * Delete pokemon
   */
  async deletePokemon(pokemon: PokemonModel): Promise<void> {
    this.remove(() => PokemonService.remove(pokemon), PokemonService.getAll);
  }

  /**
   * Mount component
   */
  public async componentDidMount(): Promise<void> {
    await this.loadPokemons();
  }
  //#endregion

  render() {
    // SETEAR DATA IN LIST
    const list = this.state.list
      .sort((a, b) => (a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1))
      .filter((p) =>
        this.state.search
          ? Object.values({ ...p, image: undefined, type: undefined })
              .join()
              .toLowerCase()
              .indexOf(this.state.search.toLowerCase()) > -1
          : true
      );
    return (
      <>
        {/* CRUD BAR */}
        <AppCrudBar
          title="Listado de Pokemon"
          onClickAdd={() =>
            this.setState({ openForm: true, element: new PokemonModel({}) })
          }
          onSearch={(search) => this.setState({ search })}
        />

        {/* TABLE */}
        <AppTable>
          <thead>
            <tr>
              <td>Nombre</td>
              <td>Imagen</td>
              <td>Ataque</td>
              <td>Defensa</td>
              <td>Acciones</td>
            </tr>
          </thead>
          <tbody>
            {list.map((row, index) => (
              <tr key={row.id}>
                <td style={{ textAlign: "left" }}>{row.name}</td>
                <td>
                  <img src={row.image} height="50"></img>
                </td>
                <td style={{ textAlign: "left" }}>{row.attack}</td>
                <td style={{ textAlign: "left" }}>{row.defense}</td>
                <td style={{ textAlign: "center" }}>
                  <i
                    className="edit"
                    onClick={() => this.toEdit(row, PokemonService.getByID)}
                  ></i>
                  <i
                    className="delete"
                    onClick={() => this.deletePokemon(row)}
                  ></i>
                </td>
              </tr>
            ))}
            {list.length === 0 && this.state.search && (
              <tr>
                <td colSpan={5}>Busqueda sin resultados</td>
              </tr>
            )}
          </tbody>
        </AppTable>

        {/* DIALOG */}
        <AppFormDialog
          open={this.state.openForm}
          title="Nuevo Pokemon"
          onClose={() => {
            this.setState({ openForm: false });
          }}
          onSubmit={() => {
            this.submit();
          }}
          disabled={
            !(
              this.state.element.name.length > 0 &&
              this.state.element.image.length > 0
            )
          }
        >
          <table>
            <tbody>
              <tr>
                <td>
                  <AppTextField
                    className="text-field"
                    label="Nombre:"
                    name="name"
                    value={this.state.element.name}
                    onChange={(v) => this.elementProp("name", v)}
                    required
                  />
                </td>
                <td>
                  <AppSliderBar
                    label="Ataque:"
                    valueMinimum={0}
                    valueMaximum={100}
                    name="attack"
                    value={this.state.element.attack}
                    onChange={(v) => this.elementProp("attack", v)}
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <AppTextField
                    className="text-field"
                    label="Imagen:"
                    placeholder="url"
                    name="image"
                    value={this.state.element.image}
                    onChange={(v) => this.elementProp("image", v)}
                    required
                    type="url"
                  />
                </td>
                <td>
                  <AppSliderBar
                    label="Defensa:"
                    valueMinimum={0}
                    valueMaximum={100}
                    name="defense"
                    value={this.state.element.defense}
                    onChange={(v) => this.elementProp("defense", v)}
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </AppFormDialog>
      </>
    );
  }
}
