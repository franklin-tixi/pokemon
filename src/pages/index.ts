/**
 * Barrel for Pokemon
 */
export * from "./PokemonModel";
export * from "./PokemonService";
export * from "./PokemonPage";
