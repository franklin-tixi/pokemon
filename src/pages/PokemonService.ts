import { PokemonModel } from "./PokemonModel";

/**
 * Class to define the services and consume the pokemon api
 */
export class PokemonService {
  /**
   * Consult all the register by an accountant
   * @returns
   */
  static async getAll(): Promise<Array<PokemonModel>> {
    return new Promise<Array<PokemonModel>>((resolve, reject) => {
      fetch("https://bp-pokemons.herokuapp.com/10?idAuthor=1")
        .then((res) => res.json())
        .then((data) => resolve(data))
        .catch((error) => reject(error));
    });
  }

  /**
   * Get data a pokemon
   * @param pokemonID
   * @returns
   */
  static async getByID(pokemonID: number): Promise<PokemonModel> {
    return new Promise<PokemonModel>((resolve, reject) => {
      fetch(`https://bp-pokemons.herokuapp.com/${pokemonID}`)
        .then((res) => res.json())
        .then((data) => resolve(new PokemonModel(data)))
        .catch((error) => reject(`No existe el pokemon id ${pokemonID}`));
    });
  }

  /**
   * create a pokemon
   * @param pokemon
   * @returns
   */
  static async create(pokemon: PokemonModel): Promise<PokemonModel> {
    return new Promise<PokemonModel>((resolve, reject) => {
      fetch("https://bp-pokemons.herokuapp.com/?idAuthor=1", {
        method: "POST",
        mode: "cors",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(pokemon),
      })
        .then((res) => res.json())
        .then((data) => resolve(data))
        .catch((error) => reject(error));
    });
  }

  /**
   * Edit a pokemon
   * @param pokemon
   * @returns
   */
  static async edit(pokemon: PokemonModel): Promise<PokemonModel> {
    return new Promise<PokemonModel>((resolve, reject) => {
      fetch(`https://bp-pokemons.herokuapp.com/${pokemon.id}`, {
        method: "PUT",
        mode: "cors",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(pokemon),
      })
        .then((res) => res.json())
        .then((data) => resolve(data))
        .catch((error) => reject(error));
    });
  }

  /**
   * Delete a pokemon
   * @param model '
   * @returns
   */
  static async remove(model: PokemonModel): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {
      fetch(`https://bp-pokemons.herokuapp.com/${model.id}`, {
        method: "DELETE",
        mode: "cors",
      })
        .then((res) => res.json())
        .then((data) => resolve(data.success))
        .catch((error) => reject(error));
    });
  }
}
