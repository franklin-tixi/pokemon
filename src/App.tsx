import "./App.css";
import { PokemonPage } from "./pages/PokemonPage";
function App() {
  return (
    <>
      <PokemonPage />
    </>
  );
}

export default App;
