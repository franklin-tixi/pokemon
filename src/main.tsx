import React from "react";
import ReactDOM from "react-dom/";
import App from "./App";
import "./index.css";
import "./components/button/AppButton.css";
import "./components/dialog/AppFormDialog.css";
import "./components/inputs/AppTextField.css";
import "./components/table/AppTable.css";

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById("root")
);
