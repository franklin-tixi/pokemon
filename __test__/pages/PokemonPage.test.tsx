import { fireEvent, render, screen } from "@testing-library/react";

import React, { Component } from "react";
import { PokemonModel, PokemonPage } from "../../src/pages";
describe("Pruebas en PokemonPage", () => {
  test("Pagina inicio Pokemon", () => {
    //@ts-ignore
    const { container } = render(<PokemonPage />);

    expect(container).toMatchSnapshot();
  });

  test("Abrir modal y crear un pokemon", () => {
    const name = "Pikachu";
    const url =
      "https://assets.stickpng.com/images/580b57fcd9996e24bc43c325.png";
    //@ts-ignore
    render(<PokemonPage />);
    fireEvent.click(screen.getByText("Nuevo"));

    const inputName = screen.getByRole("textbox", { name: "input-name" });
    const inputUrl = screen.getByRole("textbox", { name: "input-url" });
    const form = screen.getByRole("form");

    fireEvent.input(inputName, { target: { value: name } });
    fireEvent.input(inputUrl, { target: { value: url } });
    fireEvent.submit(form);
    screen.debug();
  });
});
