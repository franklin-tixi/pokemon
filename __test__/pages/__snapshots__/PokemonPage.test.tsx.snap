// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en PokemonPage Pagina inicio Pokemon 1`] = `
<div>
  <h3>
    Listado de Pokemon
  </h3>
  <div
    style="display: flex; justify-content: space-between;"
  >
    <div>
      <input
        aria-label="input-search"
        class="search text-field"
        name="search"
        placeholder="Buscar"
        style="margin-right: 1rem;"
        type="text"
        value=""
      />
    </div>
    <div>
      <button
        class="btn-new"
        style="margin-left: 1rem;"
      >
        Nuevo
      </button>
    </div>
  </div>
  <div
    style="display: flex; justify-content: center; padding: 2rem;"
  >
    <table
      class="data-table"
    >
      <thead>
        <tr>
          <td>
            Nombre
          </td>
          <td>
            Imagen
          </td>
          <td>
            Ataque
          </td>
          <td>
            Defensa
          </td>
          <td>
            Acciones
          </td>
        </tr>
      </thead>
    </table>
  </div>
</div>
`;
